### ISSUE TITLE
adwaita-qt: dark theme for Qt applications doesn’t work on NixOS 20.09

### ISSUE THREAD

**Issue by unclechu** [NixOS/nixpkgs#98009]
It used to work before 20.09 on `nixos-unstable` channel.

Here https://github.com/unclechu/nixos-config/blob/41e8bc34a12d018f645cfabcd892cab381e583c1/configuration.nix#L144-L148 I’m setting Qt theme:

``` nix
  qt5 = {
    enable = true;
    platformTheme = "gnome";
    style = "adwaita-dark";
  };
```

If I for instance run `vlc` I see this error message telling me that `adwaita-dark` is an incorrect style value:

```
VLC media player 3.0.11.1 Vetinari (revision 3.0.11.1-0-g52483f3ca2)
[0000000000649610] main libvlc: Running vlc with the default interface. Use 'cvlc' to use vlc without interface.
QApplication: invalid style override 'adwaita-dark' passed, ignoring it.
	Available styles: Windows, Fusion
```

Same story for other apps like `smplayer`, `audacious` and `psi-plus`.

**Comment by worldofpeace** [comment]
Perhaps related to https://github.com/FedoraQt/adwaita-qt/issues/140 and issues with qt https://github.com/FedoraQt/adwaita-qt/issues/140#issuecomment-684251930

**Comment by worldofpeace** [comment]
I believe this is probably related to the fact that we now use Qt 5.12 on 20.09 because that has a support period lasting for the duration of the releases (while the version in unstable doesn't) I'm thinking this commit was in the qt version in unstable https://github.com/qt/qtbase/commit/7218665699e31ac8eebab17afd684d1a56d12ac3, but with the version of qt in 20.09 it doesn't have this patch. This is just a guess because all the versions mentioned in the bug reports https://bugreports.qt.io/browse/QTBUG-86362 https://bugreports.qt.io/browse/QTBUG-85469 make me think the version in 20.09 shouldn't have this issue.

**Comment by worldofpeace** [comment]
cc @ttuegel ?

**Comment by ttuegel** [comment]
This sounds like the `qgtk3` style plugin isn't being loaded. Could you try running with `QT_DEBUG_PLUGINS=1` set in the environment?

**Comment by unclechu** [comment]
@ttuegel Here is the log of **vlc** with that env var: [vlc-qt-debug-plugins-log.txt](https://github.com/NixOS/nixpkgs/files/5224821/vlc-qt-debug-plugins-log.txt)

**Comment by gnidorah** [comment]
Same issue using builtin gtk3 platform theme: https://code.qt.io/cgit/qt/qtbase.git/tree/src/plugins/platformthemes/gtk3
This worked with Qt 5.14:
```
  environment.variables.QT_QPA_PLATFORMTHEME = "gtk3";
  environment.variables.QT_STYLE_OVERRIDE = "adwaita";
  services.xserver.windowManager.i3.extraPackages =
    options.services.xserver.windowManager.i3.extraPackages.default ++ [ pkgs.adwaita-qt ];
```
and broke with Qt 5.15:
```
QApplication: invalid style override 'adwaita' passed, ignoring it.
        Available styles: Windows, Fusion
```

**Comment by ttuegel** [comment]
Why do you need to set `QT_STYLE_OVERRIDE` at all? The QPA plugin (`qgtk3` in this case) is supposed to use the current theme. `QT_STYLE_OVERRIDE` shouldn't be set to the GTK theme name anyway; that's used to force a particular _built-in_ Qt style.

**Comment by unclechu** [comment]
@ttuegel I personally didn’t set any QT-related variables and it did work before the update.

**Comment by gnidorah** [comment]
@ttuegel 

> Why do you need to set `QT_STYLE_OVERRIDE` at all? The QPA plugin (`qgtk3` in this case) is supposed to use the current theme. `QT_STYLE_OVERRIDE` shouldn't be set to the GTK theme name anyway; that's used to force a particular _built-in_ Qt style.

AFAIK both builtin qgtk3 and external qgnomeplatform don't read theme name from gtk configs. Though, qgnomeplatform is able to read theme name from dconf database. So that's why.

qt5.nix module passes gtk theme name via `QT_STYLE_OVERRIDE` too https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/config/qt5.nix#L99 and that worked well for Qt 5.14

**Comment by unclechu** [comment]
By the way, here is how you fix it for **Qt4** (not for Qt5):
https://github.com/unclechu/nixos-config/commit/107122278d75c2a21a1b7f519e8a384b3467f270
For instance `keepassx` uses **Qt4**.

**Comment by unclechu** [comment]
According to this https://bugreports.qt.io/browse/QTBUG-85469 bug report it’s fixed in Qt `5.15.1` so maybe updating Qt version would be enough to fix this?

**Comment by unclechu** [comment]
Are there any plans to update Qt version to `5.15.1` or higher soon?

**Comment by doronbehar** [comment]
Could anyone check if https://github.com/NixOS/nixpkgs/pull/101369 fixes this? (It will take quiet some time to rebuild your system on top of it).

**Comment by unclechu** [comment]
@doronbehar It doesn’t seem it fixes anything for me:
``` sh
nix-shell -E 'let pkgs=import (fetchTarball {url="https://github.com/NixOS/nixpkgs/archive/54f7498601dcfc884c42a6d29f15475e6e1a802b.tar.gz";}) {}; in pkgs.mkShell {buildInputs = [pkgs.vlc];}' --run vlc
```
```
VLC media player 3.0.11.1 Vetinari (revision 3.0.11.1-0-g52483f3ca2)
[0000000000fba610] main libvlc: Running vlc with the default interface. Use 'cvlc' to use vlc without interface.
QApplication: invalid style override 'adwaita-dark' passed, ignoring it.
	Available styles: Windows, Fusion
[0000000001047380] main playlist: playlist is empty
QObject::~QObject: Timers cannot be stopped from another thread
```

**Comment by gnidorah** [comment]
@unclechu 
vlc is still built with 5.14
I suppose these will fix the problem
```diff
diff --git a/pkgs/development/libraries/qt-5/5.15/default.nix b/pkgs/development/libraries/qt-5/5.15/default.nix
index 1a90c610a5b..0e3bc2ae5cd 100644
--- a/pkgs/development/libraries/qt-5/5.15/default.nix
+++ b/pkgs/development/libraries/qt-5/5.15/default.nix
@@ -66,6 +66,11 @@ let
         ./qtbase.patch.d/0009-qtbase-qtpluginpath.patch
         ./qtbase.patch.d/0010-qtbase-assert.patch
         ./qtbase.patch.d/0011-fix-header_module.patch
+        (fetchpatch {
+          url = "https://codereview.qt-project.org/gitweb?p=qt/qtbase.git;a=patch;h=7218665699e31ac8eebab17afd684d1a56d12ac3";
+          name = "Ensure-styles-always-get-to-polish-the-application-palette";
+          sha256 = "1nwpds9rd7sfx7mw79d2vk3xynnqijn36mgvi7c8kyss1hp7g6iz";
+        })
       ];
     qtdeclarative = [ ./qtdeclarative.patch ];
     qtscript = [ ./qtscript.patch ];
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 1b04ab9e25b..f02bb999ae2 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -18799,7 +18799,13 @@ in
 
   adementary-theme = callPackage ../data/themes/adementary { };
 
-  adwaita-qt = libsForQt5.callPackage ../data/themes/adwaita-qt { };
+  adwaita-qt = symlinkJoin {
+    name = "adwaita-qt";
+    paths = [
+      (libsForQt5.callPackage ../data/themes/adwaita-qt { })
+      (libsForQt514.callPackage ../data/themes/adwaita-qt { })
+    ];
+  };
 
   agave = callPackage ../data/fonts/agave { };
 

```

**Comment by unclechu** [comment]
@gnidorah Any ideas how to apply patch for a derivation made by `fetchTarball`?

**Comment by doronbehar** [comment]
> @gnidorah Any ideas how to apply patch for a derivation made by `fetchTarball`?

I don't think you'd like to merge such a change to your nixpkgs fork or whatever - it will trigger rebuilds of a lot of other packages.

> vlc is still built with 5.14
> I suppose this should resolve issue for both Qt 5.14 and Qt 5.15 programs

That qt5.15.0 patch is worth trying, but please consider https://github.com/NixOS/nixpkgs/pull/102216 . 

OTH, the adwaita-qt patch is not a proper patch IMO. `adwaita-qt` should be `libsForQt5.adwaita` and `libsForQt514.adwaita` etc. Then, in the nixos module it should put both adwaita builds in the system path.

**Comment by gnidorah** [comment]
> > @gnidorah Any ideas how to apply patch for a derivation made by `fetchTarball`?
> 
> I don't think you'd like to merge such a change to your nixpkgs fork or whatever - it will trigger rebuilds of a lot of other packages.
> 
> > vlc is still built with 5.14
> > I suppose this should resolve issue for both Qt 5.14 and Qt 5.15 programs
> 
> That qt5.15.0 patch is worth trying, but please consider #102216 .

Yes, its ok to wait for  #102216 landed

> 
> OTH, the adwaita-qt patch is not a proper patch IMO. `adwaita-qt` should be `libsForQt5.adwaita` and `libsForQt514.adwaita` etc. Then, in the nixos module it should put both adwaita builds in the system path.

Yes, you're right. Same applies to other themes that provide `lib/qt-5.*/plugins/styles/*.so` like `qtcurve`

**Comment by doronbehar** [comment]
Proposed fix: https://github.com/NixOS/nixpkgs/pull/102594

**Comment by unclechu** [comment]
Now in the `nixos-20.09` release branch `vlc`, `smplayer`, `audacious`, `audacity` are all using dark theme now.
But only `psi-plus` for me is still not using dark theme. I’m not sure why but #102594 somehow fixes it.

**Comment by unclechu** [comment]
`psi-plus` is also fixed now since NixOS 21.05 release. `vlc`, `smplayer`, `audacious`, and `audacity`, all now are using dark theme. I don’t see any application now that still have this issue. Closing this one.
